src/dune_rules/include_preprocessed_sources.ml

```diff
@@ -0,0 +1,257 @@
+open Import
+open Memo.O
+
+module Lib_to_preprocess = struct
+  type t =
+    { name : string
+    ; ppx_driver_and_flags : (Path.Build.t * string list) Action_builder.t
+    ; sources : Module.t list
+    }
+
+  let ppx_driver ~sctx ~scope ~expander lib =
+    let* pps = Resolve.Memo.peek (Lib.pps lib) in
+    match pps with
+    | Ok [] | Error () -> Memo.return None
+    | Ok pps ->
+      let lib_name =
+        match Lib_name.to_local (Loc.none, Lib.name lib) with
+        | Ok lib_name -> Some lib_name
+        | Error _ -> None
+      in
+      let pps_list = List.map pps ~f:(fun lib -> Loc.none, Lib.name lib) in
+      let ctx = Super_context.context sctx in
+      let driver =
+        Preprocessing.get_ppx_driver
+          ~loc:Loc.none
+          ~expander
+          ~scope
+          ~lib_name
+          ~flags:[]
+          ctx
+          pps_list
+      in
+      Memo.return (Some driver)
+  ;;
+
+  let sources ~sctx lib =
+    let+ modules = Dir_contents.modules_of_lib sctx lib in
+    match modules with
+    | None -> []
+    | Some ms ->
+      Modules.fold_user_written ms ~init:[] ~f:(fun module_ acc -> module_ :: acc)
+  ;;
+
+  let from_lib ~sctx ~scope ~expander lib =
+    let* ppx_driver_and_flags = ppx_driver ~sctx ~scope ~expander lib in
+    match ppx_driver_and_flags with
+    | Some ppx_driver_and_flags ->
+      let+ sources = sources ~sctx lib in
+      let name = Lib_name.to_string (Lib.name lib) in
+      Some { name; ppx_driver_and_flags; sources }
+    | None -> Memo.return None
+  ;;
+end
+
+module Libs_and_exes = Monoid.Appendable_list (struct
+    type t = Lib.t
+  end)
+
+module Source_tree_map_reduce = Source_tree.Dir.Make_map_reduce (Memo) (Libs_and_exes)
+
+let libs_or_exes_under_dir ~sctx ~scope dir =
```

Comment:
What's the difference between this function and the one in utop.ml?

Reply:
I fixed the executable part and refactored a bit this function so it now differs significantly from the one in utop.